plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/TestSqlServerTypeMapping.java

```diff
@@ -69,6 +72,107 @@ public void testVarbinary()
+
+                // maximal value for a precision
+                .addRoundTrip("time(0)", "TIME '23:59:59'", createTimeType(0), "TIME '23:59:59'")
+                .addRoundTrip("time(1)", "TIME '23:59:59.9'", createTimeType(1), "TIME '23:59:59.9'")
+                .addRoundTrip("time(2)", "TIME '23:59:59.99'", createTimeType(2), "TIME '23:59:59.99'")
+                .addRoundTrip("time(3)", "TIME '23:59:59.999'", createTimeType(3), "TIME '23:59:59.999'")
+                .addRoundTrip("time(4)", "TIME '23:59:59.9999'", createTimeType(4), "TIME '23:59:59.9999'")
+                .addRoundTrip("time(5)", "TIME '23:59:59.99999'", createTimeType(5), "TIME '23:59:59.99999'")
+                .addRoundTrip("time(6)", "TIME '23:59:59.999999'", createTimeType(6), "TIME '23:59:59.999999'")
+                .addRoundTrip("time(7)", "TIME '23:59:59.9999999'", createTimeType(7), "TIME '23:59:59.9999999'")
+
+                .execute(getQueryRunner(), trinoCreateAndInsert(getSession(), "test_time"));
+
+        SqlDataTypeTest.create()
+                // round down
+                .addRoundTrip("TIME '00:00:00.00000001'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '00:00:00.000000000001'", "TIME '00:00:00.0000000'")
+
+                // round down, maximal value
+                .addRoundTrip("TIME '00:00:00.00000004'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '00:00:00.000000049'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '00:00:00.0000000449'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '00:00:00.00000004449'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '00:00:00.000000044449'", "TIME '00:00:00.0000000'")
+
+                // round up to next day, minimal value
+                .addRoundTrip("TIME '23:59:59.99999995'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.999999950'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.9999999500'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.99999995000'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.999999950000'", "TIME '00:00:00.0000000'")
+
+                // round up to next day, maximal value
+                .addRoundTrip("TIME '23:59:59.99999999'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.999999999'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.9999999999'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.99999999999'", "TIME '00:00:00.0000000'")
+                .addRoundTrip("TIME '23:59:59.999999999999'", "TIME '00:00:00.0000000'")
+
+                // round down
+                .addRoundTrip("TIME '23:59:59.999999949999'", "TIME '23:59:59.9999999'")
+
+                .execute(getQueryRunner(), trinoCreateAndInsert(getSession(), "test_time"));
+
+        SqlDataTypeTest.create()
```

Comment:
This deserves a comment IMO about the repetition and not merging with previous case.

Also you can assign the result of `SqlDataType.create` to a variable and call `execute` on it multiple times (or chain multiple `.execute` calls together.

Reply:
After limiting max precision to 9, the rounding is now done in the engine and I could collapse all blocks into one. I needed to look at this with a fresh head, thanks!